File: go/test/run_pagespeed_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/pagespeed-sdk/go"
	"github.com/voxgig-sdk/pagespeed-sdk/go/core"

	vs "github.com/voxgig-sdk/pagespeed-sdk/go/utility/struct"
)

func TestRunPagespeedEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.RunPagespeed(nil)
		if ent == nil {
			t.Fatal("expected non-nil RunPagespeedEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := run_pagespeedBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "run_pagespeed." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set PAGESPEED_TEST_RUN_PAGESPEED_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		runPagespeedRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.run_pagespeed", setup.data)))
		var runPagespeedRef01Data map[string]any
		if len(runPagespeedRef01DataRaw) > 0 {
			runPagespeedRef01Data = core.ToMapAny(runPagespeedRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = runPagespeedRef01Data

		// LOAD
		runPagespeedRef01Ent := client.RunPagespeed(nil)
		runPagespeedRef01MatchDt0 := map[string]any{
			"id": runPagespeedRef01Data["id"],
		}
		runPagespeedRef01DataDt0Loaded, err := runPagespeedRef01Ent.Load(runPagespeedRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		runPagespeedRef01DataDt0LoadResult := core.ToMapAny(runPagespeedRef01DataDt0Loaded)
		if runPagespeedRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if runPagespeedRef01DataDt0LoadResult["id"] != runPagespeedRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func run_pagespeedBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "run_pagespeed", "RunPagespeedTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read run_pagespeed test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse run_pagespeed test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"run_pagespeed01", "run_pagespeed02", "run_pagespeed03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("PAGESPEED_TEST_RUN_PAGESPEED_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"PAGESPEED_TEST_RUN_PAGESPEED_ENTID": idmap,
		"PAGESPEED_TEST_LIVE":      "FALSE",
		"PAGESPEED_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["PAGESPEED_TEST_RUN_PAGESPEED_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["PAGESPEED_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewPagespeedSDK(core.ToMapAny(mergedOpts))
	}

	live := env["PAGESPEED_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["PAGESPEED_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/run_pagespeed_entity_test.rb
# RunPagespeed entity test

require "minitest/autorun"
require "json"
require_relative "../Pagespeed_sdk"
require_relative "runner"

class RunPagespeedEntityTest < Minitest::Test
  def test_create_instance
    testsdk = PagespeedSDK.test(nil, nil)
    ent = testsdk.RunPagespeed(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = run_pagespeed_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "run_pagespeed." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set PAGESPEED_TEST_RUN_PAGESPEED_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    run_pagespeed_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.run_pagespeed")))
    run_pagespeed_ref01_data = nil
    if run_pagespeed_ref01_data_raw.length > 0
      run_pagespeed_ref01_data = Helpers.to_map(run_pagespeed_ref01_data_raw[0][1])
    end

    # LOAD
    run_pagespeed_ref01_ent = client.RunPagespeed(nil)
    run_pagespeed_ref01_match_dt0 = {
      "id" => run_pagespeed_ref01_data["id"],
    }
    run_pagespeed_ref01_data_dt0_loaded = run_pagespeed_ref01_ent.load(run_pagespeed_ref01_match_dt0, nil)
    run_pagespeed_ref01_data_dt0_load_result = Helpers.to_map(run_pagespeed_ref01_data_dt0_loaded.respond_to?(:data_get) ? run_pagespeed_ref01_data_dt0_loaded.data_get : run_pagespeed_ref01_data_dt0_loaded)
    assert !run_pagespeed_ref01_data_dt0_load_result.nil?
    assert_equal run_pagespeed_ref01_data_dt0_load_result["id"], run_pagespeed_ref01_data["id"]

  end
end

def run_pagespeed_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "run_pagespeed", "RunPagespeedTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = PagespeedSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["run_pagespeed01", "run_pagespeed02", "run_pagespeed03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["PAGESPEED_TEST_RUN_PAGESPEED_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "PAGESPEED_TEST_RUN_PAGESPEED_ENTID" => idmap,
    "PAGESPEED_TEST_LIVE" => "FALSE",
    "PAGESPEED_TEST_EXPLAIN" => "FALSE",
    "PAGESPEED_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["PAGESPEED_TEST_RUN_PAGESPEED_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["PAGESPEED_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["PAGESPEED_APIKEY"],
      },
      extra || {},
    ])
    client = PagespeedSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["PAGESPEED_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["PAGESPEED_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
